### PR TITLE
Add custom timeout for pinning requests

### PR DESCRIPTION
**Pull request by cypt4**
Resolves https://github.com/brave/brave-browser/issues/29493

Resolves 

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/api_request_helper/api_request_helper.h [hunk 1]
```diff
@@ -77,6 +78,12 @@ class APIRequestHelper {
  public:
   using Ticket = std::list<std::unique_ptr<network::SimpleURLLoader>>::iterator;
 
+  struct RequestOptions {
+    bool auto_retry_on_network_change = false;
+    size_t max_body_size = -1u;
+    absl::optional<base::TimeDelta> timeout;
+  };
```
Comment (supermassive) [inline components/api_request_helper/api_request_helper.h:85]: Lets move out of this class as api_request_helper::APIRequestOptions

File: components/ipfs/ipfs_service.cc [hunk 2]
```diff
@@ -405,9 +412,11 @@ void IpfsService::AddPin(const std::vector<std::string>& cids,
       requests_list_.insert(requests_list_.begin(), std::move(url_loader));
 
   iter->get()->Request(
-      "POST", gurl, std::string(), std::string(), false,
+      "POST", gurl, std::string(), std::string(),
       base::BindOnce(&IpfsService::OnPinAddResult, base::Unretained(this),
                      cids.size(), recursive, iter, std::move(callback)),
+      api_request_helper::APIRequestHelper::RequestOptions(false, -1u,
```
Comment (supermassive) [inline components/ipfs/ipfs_service.cc:418]: Make use of https://google.github.io/styleguide/cppguide.html#Designated_initializers 
So RequestOptions{.timeout = base::Minutes(2)}

File: components/api_request_helper/api_request_helper.cc [hunk 3]
```diff
@@ -192,6 +212,7 @@ std::unique_ptr<network::SimpleURLLoader> APIRequestHelper::CreateLoader(
   if (!payload.empty()) {
     url_loader->AttachStringForUpload(payload, payload_content_type);
   }
+  url_loader->SetTimeoutDuration(base::Seconds(30));
```
Comment (supermassive) [inline components/api_request_helper/api_request_helper.cc:215]: Do we actually need to change current behavior ?

**Review by supermassive — APPROVED** [review]
LGTM with some minor suggestions

File: components/api_request_helper/api_request_helper.cc (hunk 3 above)
Comment (cypt4) [reply components/api_request_helper/api_request_helper.cc:215]: Debug thing :facepalm:

**Comment by srirambv** [comment]
Verification passed on

Brave | 1.52.36 Chromium: 112.0.5615.49 (Official Build) nightly (64-bit)
-- | --
Revision | `bd2a7bcb881c11e8cfe3078709382934e3916914-refs/branch-heads/5615@{#936}`
OS | Windows 11 Version 22H2 (Build 22621.1555)

- Verified steps from PR
- Verified pinning is not restarted for tokens with non-ipfs metadata\image
- Verified status is retained between browser restart

https://user-images.githubusercontent.com/17010094/231439485-13eca1df-b1ba-42c4-8b93-17cb0b76ef91.mp4

**Pull request merged**